File: binary_search_recursive.py
def binary_search_recursive(arr, low, high, val):
  """
  Performs iterative binary search on a sorted array. 

  Parameters
  ----------
  arr : list, tuple or set
    A sorted linear container of integers of floats.
  
  low : int
    Lower bound of interval.

  high : int 
    Upper bound of interval.

  val : int or float
    A value to look for in arr. 
  """

  if low <= high:

      mid = (high + low) // 2

      if arr[mid] < val: 
          return binary_search(arr, mid + 1, high, val) # Ignore left half
      elif arr[mid] > val:
          return binary_search(arr, low, mid - 1, val)  # Ignore right half
      else:
          return mid                                    # Found at mid

  else:
      
      # Not found -> return default value
      return -1

File: binary_search_recursive.c
int binary_search_recursive(const int* arr,
                            const int arr_size,
                            const int low,
                            const high,
                            const int target)
{  
    /*
    Performs recursive binary search on a sorted integer array. 
    
    Parameters
    ----------
    tab : int*
        A sorted integer array.
    
    arr_size : int
        Size of input array.
        
    low : int
        Lower bound of interval.
        
    high : int
        Upper bound of interval.
    
    target : int
        Value to look for in input array.
        
    Returns the index at which the value is found or -1.
    */
    
    if (low <= high)
    {
        
        int mid = (low + high) / 2;
        
        if (arr[mid] < target)
        {
            // Ignore left half
            return binary_search_recursive(arr, 
                                           arr_size, 
                                           mid + 1, 
                                           high, 
                                           target);
        }
        else if (arr[mid] > target)
        {
            // Ignore right half
            return binary_search_recursive(arr, 
                                           arr_size, 
                                           low, 
                                           mid - 1, 
                                           target);
        }
        else
        {
            // Found at mid
            return mid;    
        }
    }
    else
    {
        // Not found -> return default value
        return -1;
    }
    
}
